File: src/leader_board/protocol.py
# Chunk: docs/chunks/leader_board_local_server - Local WebSocket server adapter
"""Wire protocol frame parsing and serialization.

Defines typed dataclasses for every frame in the leader board wire protocol
(JSON over WebSocket). This module is shared between the local server adapter
and the Durable Objects adapter — both speak the same wire format.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Union


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class InvalidFrameError(Exception):
    """Raised when a frame cannot be parsed (malformed JSON, unknown type, missing fields)."""

    pass


# ---------------------------------------------------------------------------
# Client → Server frames
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class AuthFrame:
    """Client response to a challenge (existing swarm)."""

    swarm: str
    signature: str  # hex-encoded Ed25519 signature


@dataclass(frozen=True)
class RegisterSwarmFrame:
    """Client request to register a new swarm."""

    swarm: str
    public_key: str  # hex-encoded Ed25519 public key


@dataclass(frozen=True)
class WatchFrame:
    """Subscribe to a channel starting after a cursor position."""

    channel: str
    swarm: str
    cursor: int


@dataclass(frozen=True)
class SendFrame:
    """Append an encrypted message to a channel."""

    channel: str
    swarm: str
    body: str  # base64-encoded ciphertext


@dataclass(frozen=True)
class ChannelsFrame:
    """List all channels in a swarm."""

    swarm: str


@dataclass(frozen=True)
class SwarmInfoFrame:
    """Retrieve metadata about a swarm."""

    swarm: str


ClientFrame = Union[
    AuthFrame, RegisterSwarmFrame, WatchFrame, SendFrame, ChannelsFrame, SwarmInfoFrame
]


# ---------------------------------------------------------------------------
# Server → Client frames
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class ChallengeFrame:
    """Server challenge sent on connection open."""

    nonce: str  # hex-encoded 32-byte nonce


@dataclass(frozen=True)
class AuthOkFrame:
    """Server confirmation of successful authentication."""

    pass


@dataclass(frozen=True)
class MessageFrame:
    """A single message delivered in response to a watch."""

    channel: str
    position: int
    body: str  # base64-encoded ciphertext
    sent_at: str  # ISO 8601 UTC


@dataclass(frozen=True)
class AckFrame:
    """Confirmation that a send was appended."""

    channel: str
    position: int


@dataclass(frozen=True)
class ChannelsListFrame:
    """Response listing all channels in a swarm."""

    channels: list[dict]  # [{"name": ..., "head_position": ..., "oldest_position": ...}]


@dataclass(frozen=True)
class SwarmInfoResponseFrame:
    """Response with swarm metadata."""

    swarm: str
    created_at: str  # ISO 8601 UTC


@dataclass(frozen=True)
class PingFrame:
    """Server keepalive ping. Clients should ignore this frame."""

    pass


@dataclass(frozen=True)
class ErrorFrame:
    """Error response."""

    code: str
    message: str
    # Optional additional fields depending on error code
    earliest_position: int | None = None


ServerFrame = Union[
    ChallengeFrame,
    AuthOkFrame,
    MessageFrame,
    AckFrame,
    ChannelsListFrame,
    SwarmInfoResponseFrame,
    PingFrame,
    ErrorFrame,
]


# ---------------------------------------------------------------------------
# Parsing (JSON string → ClientFrame)
# ---------------------------------------------------------------------------

_CLIENT_FRAME_REQUIRED_FIELDS: dict[str, list[str]] = {
    "auth": ["swarm", "signature"],
    "register_swarm": ["swarm", "public_key"],
    "watch": ["channel", "swarm", "cursor"],
    "send": ["channel", "swarm", "body"],
    "channels": ["swarm"],
    "swarm_info": ["swarm"],
}


def parse_client_frame(data: str) -> ClientFrame:
    """Parse a JSON string into a typed client frame.

    Raises :class:`InvalidFrameError` for malformed JSON, unknown frame types,
    or missing required fields.
    """
    try:
        obj = json.loads(data)
    except (json.JSONDecodeError, TypeError) as exc:
        raise InvalidFrameError(f"Malformed JSON: {exc}") from exc

    if not isinstance(obj, dict):
        raise InvalidFrameError("Frame must be a JSON object")

    frame_type = obj.get("type")
    if frame_type is None:
        raise InvalidFrameError("Missing 'type' field")

    required = _CLIENT_FRAME_REQUIRED_FIELDS.get(frame_type)
    if required is None:
        raise InvalidFrameError(f"Unknown frame type: {frame_type!r}")

    for field in required:
        if field not in obj:
            raise InvalidFrameError(
                f"Missing required field {field!r} for frame type {frame_type!r}"
            )

    if frame_type == "auth":
        return AuthFrame(swarm=obj["swarm"], signature=obj["signature"])
    elif frame_type == "register_swarm":
        return RegisterSwarmFrame(swarm=obj["swarm"], public_key=obj["public_key"])
    elif frame_type == "watch":
        return WatchFrame(
            channel=obj["channel"], swarm=obj["swarm"], cursor=int(obj["cursor"])
        )
    elif frame_type == "send":
        return SendFrame(
            channel=obj["channel"], swarm=obj["swarm"], body=obj["body"]
        )
    elif frame_type == "channels":
        return ChannelsFrame(swarm=obj["swarm"])
    elif frame_type == "swarm_info":
        return SwarmInfoFrame(swarm=obj["swarm"])
    else:
        raise InvalidFrameError(f"Unknown frame type: {frame_type!r}")


# ---------------------------------------------------------------------------
# Serialization (ServerFrame → JSON string)
# ---------------------------------------------------------------------------


def serialize_server_frame(frame: ServerFrame) -> str:
    """Serialize a server frame to a JSON string."""
    if isinstance(frame, ChallengeFrame):
        obj = {"type": "challenge", "nonce": frame.nonce}
    elif isinstance(frame, AuthOkFrame):
        obj = {"type": "auth_ok"}
    elif isinstance(frame, MessageFrame):
        obj = {
            "type": "message",
            "channel": frame.channel,
            "position": frame.position,
            "body": frame.body,
            "sent_at": frame.sent_at,
        }
    elif isinstance(frame, AckFrame):
        obj = {"type": "ack", "channel": frame.channel, "position": frame.position}
    elif isinstance(frame, ChannelsListFrame):
        obj = {"type": "channels_list", "channels": frame.channels}
    elif isinstance(frame, SwarmInfoResponseFrame):
        obj = {
            "type": "swarm_info",
            "swarm": frame.swarm,
            "created_at": frame.created_at,
        }
    elif isinstance(frame, PingFrame):
        obj = {"type": "ping"}
    elif isinstance(frame, ErrorFrame):
        obj: dict = {"type": "error", "code": frame.code, "message": frame.message}
        if frame.earliest_position is not None:
            obj["earliest_position"] = frame.earliest_position
    else:
        raise TypeError(f"Unknown server frame type: {type(frame)}")

    return json.dumps(obj, separators=(",", ":"))

File: workers/leader-board/src/protocol.ts
// Chunk: docs/chunks/leader_board_durable_objects - Cloudflare DO adapter
/**
 * Wire protocol types and validation for the leader board WebSocket protocol.
 *
 * Frame types match SPEC.md Wire Protocol section exactly. Clients cannot
 * distinguish between the DO adapter and the local server adapter.
 */

// --- Constants ---

const CHANNEL_NAME_RE = /^[a-zA-Z0-9_-]{1,128}$/;
const MESSAGE_MAX_BYTES = 1_048_576; // 1 MB

// --- Handshake Frames ---

export interface ChallengeFrame {
  type: "challenge";
  nonce: string;
}

export interface AuthFrame {
  type: "auth";
  swarm: string;
  signature: string;
}

export interface RegisterSwarmFrame {
  type: "register_swarm";
  swarm: string;
  public_key: string;
}

export interface AuthOkFrame {
  type: "auth_ok";
}

// --- Client → Server Frames (Post-Auth) ---

export interface WatchFrame {
  type: "watch";
  channel: string;
  swarm: string;
  cursor: number;
}

export interface SendFrame {
  type: "send";
  channel: string;
  swarm: string;
  body: string;
}

export interface ChannelsFrame {
  type: "channels";
  swarm: string;
}

export interface SwarmInfoFrame {
  type: "swarm_info";
  swarm: string;
}

// --- Server → Client Frames ---

export interface MessageFrame {
  type: "message";
  channel: string;
  position: number;
  body: string;
  sent_at: string;
}

export interface AckFrame {
  type: "ack";
  channel: string;
  position: number;
}

export interface ChannelsListFrame {
  type: "channels_list";
  channels: Array<{ name: string; head_position: number; oldest_position: number }>;
}

export interface SwarmInfoResponseFrame {
  type: "swarm_info";
  swarm: string;
  created_at: string;
}

export interface ErrorFrame {
  type: "error";
  code: string;
  message: string;
  earliest_position?: number;
}

// --- Discriminated Unions ---

export type HandshakeClientFrame = AuthFrame | RegisterSwarmFrame;
export type PostAuthClientFrame = WatchFrame | SendFrame | ChannelsFrame | SwarmInfoFrame;
export type ClientFrame = HandshakeClientFrame | PostAuthClientFrame;
export type ServerFrame =
  | ChallengeFrame
  | AuthOkFrame
  | MessageFrame
  | AckFrame
  | ChannelsListFrame
  | SwarmInfoResponseFrame
  | ErrorFrame;

// --- Parsing ---

export class ProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ProtocolError";
  }
}

function requireString(obj: Record<string, unknown>, field: string): string {
  const val = obj[field];
  if (typeof val !== "string" || val.length === 0) {
    throw new ProtocolError(`Missing or invalid field: ${field}`);
  }
  return val;
}

function requireNonNegativeInt(obj: Record<string, unknown>, field: string): number {
  const val = obj[field];
  if (typeof val !== "number" || !Number.isInteger(val) || val < 0) {
    throw new ProtocolError(`Missing or invalid field: ${field} (must be a non-negative integer)`);
  }
  return val;
}

function validateChannelName(name: string): void {
  if (!CHANNEL_NAME_RE.test(name)) {
    throw new ProtocolError(
      `Invalid channel name: "${name}" (must match ${CHANNEL_NAME_RE.source})`
    );
  }
}

/**
 * Parse a raw JSON string into a typed handshake client frame.
 * Throws ProtocolError on invalid input.
 */
export function parseHandshakeFrame(raw: string): HandshakeClientFrame {
  let obj: Record<string, unknown>;
  try {
    obj = JSON.parse(raw);
  } catch {
    throw new ProtocolError("Invalid JSON");
  }

  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
    throw new ProtocolError("Frame must be a JSON object");
  }

  const type = obj.type;

  switch (type) {
    case "auth":
      return {
        type: "auth",
        swarm: requireString(obj, "swarm"),
        signature: requireString(obj, "signature"),
      };
    case "register_swarm":
      return {
        type: "register_swarm",
        swarm: requireString(obj, "swarm"),
        public_key: requireString(obj, "public_key"),
      };
    default:
      throw new ProtocolError(`Unexpected handshake frame type: ${String(type)}`);
  }
}

/**
 * Parse a raw JSON string into a typed post-auth client frame.
 * Throws ProtocolError on invalid input.
 */
export function parsePostAuthFrame(raw: string): PostAuthClientFrame {
  let obj: Record<string, unknown>;
  try {
    obj = JSON.parse(raw);
  } catch {
    throw new ProtocolError("Invalid JSON");
  }

  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
    throw new ProtocolError("Frame must be a JSON object");
  }

  const type = obj.type;

  switch (type) {
    case "watch": {
      const channel = requireString(obj, "channel");
      validateChannelName(channel);
      return {
        type: "watch",
        channel,
        swarm: requireString(obj, "swarm"),
        cursor: requireNonNegativeInt(obj, "cursor"),
      };
    }
    case "send": {
      const channel = requireString(obj, "channel");
      validateChannelName(channel);
      const body = requireString(obj, "body");
      // Validate body size (base64 encoded — decode to check actual size)
      const decodedSize = Math.ceil((body.length * 3) / 4);
      if (decodedSize > MESSAGE_MAX_BYTES) {
        throw new ProtocolError(
          `Message body too large: ~${decodedSize} bytes (max ${MESSAGE_MAX_BYTES})`
        );
      }
      return {
        type: "send",
        channel,
        swarm: requireString(obj, "swarm"),
        body,
      };
    }
    case "channels":
      return {
        type: "channels",
        swarm: requireString(obj, "swarm"),
      };
    case "swarm_info":
      return {
        type: "swarm_info",
        swarm: requireString(obj, "swarm"),
      };
    default:
      throw new ProtocolError(`Unknown frame type: ${String(type)}`);
  }
}

/** Serialize a server frame to JSON string. */
export function serializeFrame(frame: ServerFrame): string {
  return JSON.stringify(frame);
}
